x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/EventEmitterService.java

```diff
@@ -0,0 +1,104 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.entsearch.analytics;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.Marker;
+import org.apache.logging.log4j.MarkerManager;
+
+import org.elasticsearch.ResourceNotFoundException;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.get.GetRequest;
+import org.elasticsearch.action.index.IndexResponse;
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
+import org.elasticsearch.client.internal.Client;
+import org.elasticsearch.client.internal.OriginSettingClient;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.io.stream.NamedWriteableRegistry;
+import org.elasticsearch.common.io.stream.ReleasableBytesStreamOutput;
+
+import org.elasticsearch.common.util.BigArrays;
+
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentFactory;
+
+
+import java.time.format.DateTimeFormatter;
+import java.time.LocalDateTime;
+
+import static org.elasticsearch.xpack.core.ClientHelper.ENT_SEARCH_ORIGIN;
+
+public class EventEmitterService {
+    private static final Logger logger = LogManager.getLogger(EventEmitterService.class);
+    private static final Marker AUDIT_MARKER = MarkerManager.getMarker("org.elasticsearch.xpack.entsearch.analytics");
+    public static final String ANALYTICS_DATASTREAM_PATTERN = ".events-";
+
+    private final Client clientWithOrigin;
+    private final ClusterService clusterService;
+    public final NamedWriteableRegistry namedWriteableRegistry;
+    private final BigArrays bigArrays;
+
+    public static final ParseField EVENT_DATASET_NAME_FIELD = new ParseField("event.dataset");
+    public static final ParseField TIMESTAMP_FIELD = new ParseField("@timestamp");
+
+    public EventEmitterService(
+        Client client,
+        ClusterService clusterService,
+        NamedWriteableRegistry namedWriteableRegistry,
+        BigArrays bigArrays
+    ) {
+        this.clientWithOrigin = new OriginSettingClient(client, ENT_SEARCH_ORIGIN);
+        this.clusterService = clusterService;
+        this.namedWriteableRegistry = namedWriteableRegistry;
+        this.bigArrays = bigArrays;
+    }
+
+    public void emitEvent(AnalyticsEvent event, ActionListener<AcknowledgedResponse> listener) {
+        final String collectionName = getAnalyticsEventCollectionName(event);
+        final String collectionDataStream = getCollectionDataStream(collectionName);
+        final GetRequest getRequest = new GetRequest(collectionDataStream).realtime(true);
+
+        clientWithOrigin.get(getRequest, listener.delegateFailure((delegate, getResponse) -> {
```

Comment:
Are you checking if the data stream exists? This should be done through the cluster state, no need to perform a get request here.

Reply:
Yes, I was going to use the latest change from @afoucret PR